publish_process.js

```diff
@@ -0,0 +1,315 @@
+    if(dafnyReleases[i].tag_name != "nightly") {
+      mostRecentDafnyRelease = dafnyReleases[i].tag_name;
+      break;
+    }
+  }
+  if(mostRecentDafnyRelease == null) {
+    console.log(`Could not fetch the latest Dafny release version from ${dafnyReleasesURL}`);
+    throw ABORTED;
+  }
+  return mostRecentDafnyRelease;
+}
+
+
+async function readPackageJson() {
+  const packageContent = await fs.promises.readFile(packageFile);
+  const packageObj = JSON.parse(packageContent);
+  return packageObj;
+}
+
+async function writePackage(packageObj) {
+  await fs.promises.writeFile(packageFile, JSON.stringify(packageObj, null, 2));
+}
+
+// returns the new version number
+async function nextVersion(currentVersion) {
+  const currentVersionRegex = /(\d+)\.(\d+)\.(\d+)/;
+  const match = currentVersionRegex.exec(currentVersion);
+  if(match == null) {
+    console.log(`Could not parse version ${currentVersion}`);
+    throw ABORTED;
+  }
+  var bumpedPatch = `${match[1]}.${match[2]}.${parseInt(match[3]) + 1}`;
+  var bumpedMinor = `${match[1]}.${parseInt(match[2]) + 1}.0`;
+  var bumpedMajor = `${parseInt(match[1]) + 1}.0.0`;
+  console.log("Should the next version be:");
+  console.log(`${currentVersion} => ${bumpedPatch}? ${ACCEPT_HINT}`);
+  console.log(`${currentVersion} => ${bumpedMinor}? (m)`);
+  var answer = await question(`${currentVersion} => ${bumpedMajor}? (M)\n`);
+  if(ok(answer)) {
+    return bumpedPatch;
+  } else if(answer == "m") {
+    return bumpedMinor;
+  } else if(answer == "M") {
+    return bumpedMajor;
+  } else {
+    console.log("Publishing script aborted.");
+    throw ABORTED;
+  }
+}
+
+async function getLastPreparedTag() {
+  const lastCommitMessage = (await execAsync("git log -1 --pretty=%B")).stdout.trim();
+  const lastCommitMessageRegex = /(v\d+\.\d+\.\d+)/;
+  const match = lastCommitMessageRegex.exec(lastCommitMessage);
+  if(match == null) {
+    return false;
+  }
+  return match[1];
+}
+
+function getCommandLine() {
+  switch (process.platform) { 
+     case 'darwin' : return 'open';
+     case 'win32' : return 'start';
+     case 'win64' : return 'start';
+     default : return 'xdg-open';
+  }
+}
+
+async function getAllRecentCommitMessagesFormatted(currentChangeLogVersion) {
+  // Query git for all the commit messages from currentChangeLogVersion (excluded) to the latest commit (included)
+  var raw = (await execAsync(`git log --pretty=%B v${currentChangeLogVersion}..HEAD`)).stdout.trim();
+  raw = "- " + raw.trim().replace(/\n\s*/g, "\n- ");
+  raw = raw.replace(/\(#(\d+)\)/g, "(https://github.com/dafny-lang/ide-vscode/pull/$1)");
+  raw = raw.trim();
+  raw = raw.replace(/\n.*$/, ""); // Removes the last one, because it's the last release commit.
+  return raw.trim();
+}
+
+function close() {
+  rl.close();
+  return false;
+}
+
+async function Main() {
+  try {
+    // verify that we are on the master branch.
+    await ensureMaster();
```

Comment:
It's generally safer to perform a release process from a completely fresh git clone, so you aren't in danger of polluting it from an unclean state.

Reply:
How  could I pollute it from an unclean state? Anyway, I explicitly add the things I want to push so I don't think it's a concern.

Reply:
You could have locally staged files that get accidentally committed, for example.

Normally I'd recommend a fresh clone, but since the script you want to run is in the repo itself that's not ideal. I'd suggest running a couple of commands like `git diff` to ensure you're in a pristine state.